### PR TITLE
add-skill and delete-skill command added (index argument)

### PR DESCRIPTION
**Pull request by tzaph**
Tests for the command itself are not ready, but there are simple tests for the parser. The command is not final as it requires the hashtag notation.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddSkillCommand.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            + "Parameters: INDEX (must be a positive integer) "
+            + PREFIX_SKILL + " SKILL "
+            + "[" + PREFIX_SKILL + " SKILL]...\n"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_SKILL + " Python "
+            + PREFIX_SKILL + " Java";
+
+    public static final String MESSAGE_SUCCESS = "New skills added";
+    public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
+    public static final String MESSAGE_DUPLICATE_COURSE_MATE = "This courseMate already exists in the contact list";
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/commands/AddSkillCommand.java:41]: To be solved later: I think all of us are using different capitalizations of coursemate. Need to be consistent.

File: src/main/java/seedu/address/logic/commands/AddSkillCommand.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        this.queryableCourseMate = queryableCourseMate;
+        this.addSkillDescriptor = new AddSkillDescriptor(addSkillDescriptor);
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        List<CourseMate> lastShownList = model.getFilteredCourseMateList();
+
+        if (queryableCourseMate.isIndex() && queryableCourseMate.getIndex().getZeroBased() >= lastShownList.size()) {
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/commands/AddSkillCommand.java:62]: Recommended to break this into two lines.

File: src/main/java/seedu/address/logic/commands/AddSkillCommand.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        return new CommandResult(MESSAGE_SUCCESS, false, false, true);
+    }
+
+    /**
+     * Creates and returns a {@code CourseMate} with the details of {@code courseMateToEdit}
+     * edited with {@code addSkillDescriptor}.
+     */
+    private static CourseMate addSkillToCourseMate(CourseMate courseMateToEdit,
+                                                     AddSkillDescriptor addSkillDescriptor) {
+        assert courseMateToEdit != null;
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/commands/AddSkillCommand.java:91]: there is a `requireNonNull` function. Use that to retain consistency.

File: src/main/java/seedu/address/logic/commands/DeleteSkillCommand.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            + "Parameters: INDEX (must be a positive integer) "
+            + PREFIX_SKILL + " SKILL "
+            + "[" + PREFIX_SKILL + " SKILL]...\n"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_SKILL + " Python "
+            + PREFIX_SKILL + " Java";
+
+    public static final String MESSAGE_SUCCESS = "Skills are successfully deleted";
+    public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
+    public static final String MESSAGE_DUPLICATE_COURSE_MATE = "This courseMate already exists in the contact list";
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/commands/DeleteSkillCommand.java:41]: Maybe use "this name already exists in the contact list"? This message might be confusing.

File: src/main/java/seedu/address/logic/commands/DeleteSkillCommand.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    public static final String MESSAGE_SKILL_NOT_PRESENT = "This courseMate does not have one of the skills provided.";
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/commands/DeleteSkillCommand.java:42]: Recommended to state which skill is not provided (more informative messages).

File: src/main/java/seedu/address/logic/commands/DeleteSkillCommand.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        return true;
+    }
+
+    /**
+     * Creates and returns a {@code CourseMate} with the details of {@code courseMateToEdit}
+     * edited with {@code deleteSkillDescriptor}.
+     */
+    private static CourseMate deleteCourseMateSkills(CourseMate courseMateToEdit,
+                                                     DeleteSkillDescriptor deleteSkillDescriptor) {
+        assert courseMateToEdit != null;
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/commands/DeleteSkillCommand.java:112]: Same here, use `requireNonNull`.

File: src/main/java/seedu/address/logic/parser/AddSkillCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+     * If {@code skills} contain only one element which is an empty string, it will be parsed into a
+     * {@code Set<Skill>} containing zero skills.
+     */
+    private Optional<Set<Skill>> parseSkillsForEdit(Collection<String> skills) throws ParseException {
+        assert skills != null;
+
+        if (skills.isEmpty()) {
+            return Optional.empty();
+        }
+        Collection<String> skillSet = skills.size() == 1 && skills.contains("") ? Collections.emptySet() : skills;
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/parser/AddSkillCommandParser.java:63]: This line might be hard to follow. Consider using an explicit `if`.

File: src/main/java/seedu/address/logic/parser/MatchMateParser.java [hunk 8]
```diff
@@ -105,8 +113,8 @@ public Command parseCommand(String userInput) throws ParseException {
      * Checks for prefixes if they are valid with the currently supported prefix codes. The argument is valid
      * if every prefix occurence in {@code argsString} is lead and trailed by a whitespace.
      *
-     * @param argsString      Arguments string of the form: {@code preamble <prefix> value <prefix> value ...}
-     * @return                Boolean object that indicates whether {@code argsString} is valid.
+     * @param argsString Arguments string of the form: {@code preamble <prefix> value <prefix> value ...}
+     * @return Boolean object that indicates whether {@code argsString} is valid.
```
Comment (benson1029) [inline src/main/java/seedu/address/logic/parser/MatchMateParser.java:117]: The return type is already in the code, should omit `Boolean object that indicates` since it's obvious.

**Review by benson1029 — APPROVED** [review]
Some minor code quality suggestions given, need not be fixed now. The abstraction is neat overall, so good job!

**Comment by tzaph** [comment]
This also closes #17 (user story related to add-skill and delete-skill)
